File: Forecasting_Credit_Card_Defaults.R
# R Code
# Import Libraries
library(caret)
library(dplyr)
library(pROC)
library(ROCR)

credit1 = read.csv("creditcard.csv")
str(credit1)
summary(credit1)

missing = credit1 %>% summarise_all(funs(sum(is.na(.))/n()))
missing

# Let's remove the ID variable it is not needed
credit=credit1[,-1]

# There are some categories with very few observations. Let's group them together.
credit$EDUCATION[credit$EDUCATION == 0] = 4
credit$EDUCATION[credit$EDUCATION == 5] = 4
credit$EDUCATION[credit$EDUCATION == 6] = 4
credit$MARRIAGE[credit$MARRIAGE == 0] = 3

# Make some numeric variables into factors
credit$default.payment.next.month=as.factor(credit$default.payment.next.month)
credit$SEX=as.factor(credit$SEX) 
credit$EDUCATION=as.factor(credit$EDUCATION)
credit$MARRIAGE=as.factor(credit$MARRIAGE)

# Logistic model
logit.1 = glm(default.payment.next.month~., data=credit, family="binomial")
summary(logit.1)

pred = predict(logit.1, type = "response", credit)
summary(pred)

# ROC and AUC
glm.roc = roc(response = credit$default.payment.next.month, predictor = pred)
plot(glm.roc, legacy.axes = TRUE, print.auc.y = 1.0, print.auc = TRUE)
coords(glm.roc, "best", "threshold")

# Threshold = 0.5
pred_default = factor(ifelse(pred >=0.5, "Yes", "No"))
credit$default.payment.next.month = factor(ifelse(credit$default.payment.next.month==1, "Yes","No"))
confusionMatrix(pred_default, credit$default.payment.next.month, positive = "Yes")

glm.roc2 = prediction(pred, credit$default.payment.next.month)
glm.roc2a = performance(glm.roc2, "tpr", "fpr")

plot(glm.roc2a, colorize=TRUE, print.cutoffs.at=seq(0,1,by=0.1), text.adj=c(-0.2,1.7))

## Cut-off point seems to be better between 0.2 and 0.3 than in 0.5

File: Forecasting_Credit_Card_Defaults.py
# Python Code
# import modules
import pandas as pd
import os
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,classification_report, precision_recall_curve, confusion_matrix
from sklearn.model_selection import train_test_split

# Import data
credit = pd.read_csv("creditcard.csv")
print(credit)

# data summary
credit.info()
credit.describe()

# removing ID variable
credit1 = credit.drop("ID",axis=1)  # axis=0 --> row; axis=1 --> column
print(credit1)
credit1.head()

# checking for missing values
missing_values = credit1.isnull().sum()
print(missing_values)
# no missing values in the data set

# data cleaning, merging categorical variables with few observations
print(credit1)
# recording new education column
credit1['EDUCATION'] = credit1['EDUCATION'].replace({0:4, 5:4, 6:4})
# recording new marriage column
credit1['MARRIAGE'] = credit1['MARRIAGE'].replace({0:3})
print(credit1)

# making some numeric variables into factors
credit1.info()
# converting 'default.payment.next.month', 'SEX', 'EDUCATION', and 'MARRIAGE' into factors
credit1['default.payment.next.month'] = credit1['default.payment.next.month'].astype('category')
credit1['SEX'] = credit1['SEX'].astype('category')
credit1['EDUCATION'] = credit1['EDUCATION'].astype('category')
credit1['MARRIAGE'] = credit1['MARRIAGE'].astype('category')
# check
credit1.info()

# splitting the data set
credit1.info()
x = credit1[['PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6']]
y = credit1['default.payment.next.month']

# splitting the data into training and test sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

# Running a Logistical Model
model = LogisticRegression()

# fit the model to the training data
model.fit(x_train, y_train)

# Making predictions on the test data
y_pred = model.predict(x_test)
y_pred

# confusion matrix, for analyzing how well the model performed
# Print accuracy score
print("Accuracy Score:", accuracy_score(y_test, y_pred))

# Print classification report
print("\nClassification Report:\n", classification_report(y_test, y_pred))

